[SG-649] Add confirmation dialog and tweak shared key retrieval

differsthecat opened the pull request:
## Type of change

<!-- (mark with an `X`) -->

```
- [x] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Add a confirmation dialog when completing a handshake with the DuckDuckGoBrowser

## Code changes

### Test Runner
- **apps/desktop/native-messaging-test-runner/src/bw-handshake.ts:** Add message to handle canceled response in test.
- **apps/desktop/native-messaging-test-runner/src/ipcService.ts:** Increase timeout for messages slightly to account for having to accept connection.
- **apps/desktop/native-messaging-test-runner/src/nativeMessageService.ts:** Fix status return

### Desktop
- **apps/desktop/src/app/accounts/settings.component.ts:** Set DDG Shared key to null if disabling setting
- **apps/desktop/src/app/services/services.module.ts:** Add i18n Service
- **apps/desktop/src/locales/en/messages.json:** Add new messages for dialog
- **apps/desktop/src/models/nativeMessaging/unencryptedMessageResponse.ts:** Fix status responses
- **apps/desktop/src/services/nativeMessageHandler.service.ts:** Most of the changes are here
    - Add confirmation dialog
    - Save secret to storage not encrypted, but in secure storage still (I was encrypting it, but there's no way for us to decrypt without the private key which DDG will have)
    - Get stored secret when decrypting payload and send cannot-decrypt back if it doesn't exist


## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

differsthecat commented:
@danielleflinn Can I have your input on the dialog copy here for the confirmation? I'm sure it can be bette!

differsthecat commented:
@Hinton In this PR are some changes to the key like I was talking about. Some question I had are:

- Should we be calling state service here or should I move the key getting/setting to maybe crypto service? I know with ManifestV3 work we are removing the direct dependency on stater service, but wasn't sure the best place to put this.
- Is this general handling of it correct with getting it on decrypting of the payload?

Hinton commented on apps/desktop/native-messaging-test-runner/src/ipcService.ts:
```diff
@@ -13,7 +13,7 @@ NodeIPC.config.maxRetries = 0;
 NodeIPC.config.silent = true;
 
 const DESKTOP_APP_PATH = `${homedir}/tmp/app.bitwarden`;
-const DEFAULT_MESSAGE_TIMEOUT = 10 * 1000; // 10 seconds
```
Any reason to change the timeout?

Hinton replied:
Ah, makes sense. Yea or if there is some way to override the timeout just for the confirmation perhaps? To avoid the tests taking a long time before failing.

differsthecat replied:
That's a good point, I'll look into seeing if there is a reasonable way to do that!

differsthecat replied:
Added an optional `options` object to the IPC service with a timeout override that is then used by the handshake command. This seemed like the best approach and allows us to extend the options in the future for other purposes.

Hinton commented on apps/desktop/native-messaging-test-runner/src/nativeMessageService.ts:
```diff
@@ -22,7 +22,8 @@ import IPCService from "./ipcService";
 
 type HandshakePayload = {
-  status: "success" | "canceled";
+  status?: "success";
```
Boolean here.

Hinton commented on apps/desktop/native-messaging-test-runner/src/nativeMessageService.ts:
```diff
@@ -22,7 +22,8 @@ import IPCService from "./ipcService";
 type HandshakePayload = {
-  status: "success" | "canceled";
+  status?: "success";
+  error?: "canceled";
```
Any point in having an error field that can only be `canceled`? Perhaps change it to `string`? Or should we define a new enum for errors?

Hinton reviewed (commented):
We need to figure out a way to break up `nativeMessageHandler` in the future, I think it should just forward whatever command it gets to external commands that performs the actual action.

DXJ999empireZ commented on apps/desktop/src/services/nativeMessageHandler.service.ts:
```diff
@@ -91,20 +92,56 @@ export class NativeMessageHandler {
+          },
+        });
+        return;
+      }
```
https://www.bestegg.com/wp-content/uploads/2019/01/equal-housing-lender.png

differsthecat commented:
@Hinton I fixed the typings and added the ability to override the default message timeout in the test runner.

As far as extracting the calls to state service for the relevant things, the only time one of these settings is used outside of this native messaging service is in the settings service; I'm not sure what the plan is to extract all of those out, but besides this service itself, the only one I can think that would be a good owner of the shared key getting and setting would be crypto service, besides leaving them in this service.
